src/protocol/ipa/mod.rs

```diff
@@ -760,158 +762,141 @@ pub mod tests {
-                    continue;
+        for per_user_cap in [1, 3] {
+            println!("Running with a per-user-cap of {per_user_cap}.");
+
+            let mut rng = StdRng::seed_from_u64(random_seed);
+
+            let mut raw_data = Vec::with_capacity(NUM_USERS * MAX_RECORDS_PER_USER);
+            let mut expected_results = vec![0_u32; MAX_BREAKDOWN_KEY];
+
+            for _ in 0..NUM_USERS {
+                let random_user_id = rng.gen_range(0..MAX_USER_ID);
+                let num_records_for_user = rng.gen_range(1..MAX_RECORDS_PER_USER);
+                let mut records_for_user = Vec::with_capacity(num_records_for_user);
+                for _ in 0..num_records_for_user {
+                    let random_timestamp = rng.gen_range(0..SECONDS_IN_EPOCH);
+                    let is_trigger_report = rng.gen::<bool>();
+                    let random_breakdown_key = if is_trigger_report {
+                        0
+                    } else {
+                        rng.gen_range(0..MAX_BREAKDOWN_KEY)
+                    };
+                    let trigger_value = if is_trigger_report {
+                        rng.gen_range(1..MAX_TRIGGER_VALUE)
+                    } else {
+                        0
+                    };
+                    records_for_user.push(TestRawDataRecord {
+                        user_id: random_user_id,
+                        timestamp: random_timestamp,
+                        is_trigger_report,
+                        breakdown_key: random_breakdown_key,
+                        trigger_value,
+                    });
                 }
 
-                if record.is_trigger_report {
-                    encountered_trigger = true;
-                } else if encountered_trigger {
-                    expected_results[record.breakdown_key] += 1;
-                    contributed_once = true;
-                    encountered_trigger = false;
+                // sort in reverse time order
+                records_for_user.sort_unstable_by(|a, b| b.timestamp.cmp(&a.timestamp));
+
+                let mut pending_trigger_value = 0;
+                let mut total_contribution = 0;
+                for record in &records_for_user {
+                    if total_contribution >= per_user_cap {
+                        continue;
+                    }
+
+                    if record.is_trigger_report {
+                        pending_trigger_value += record.trigger_value;
+                    } else if pending_trigger_value > 0 {
+                        let delta_to_per_user_cap = per_user_cap - total_contribution;
+                        let capped_contribution =
+                            std::cmp::min(delta_to_per_user_cap, pending_trigger_value);
+                        expected_results[record.breakdown_key] += capped_contribution;
+                        total_contribution += capped_contribution;
+                        pending_trigger_value = 0;
+                    }
                 }
```

Comment:
This bit where you add to the expectation looks like another useful function to break out. It can take `records_for_user: &[...], expected_results: &mut Vec<u32>`.